src/framework/execution/__tests__/logger.test.ts

```diff
@@ -431,7 +432,102 @@ describe('createErrorEventDescription', () => {
       'testing',
     );
     expect(description).toEqual(
-      `testing (errorCode=${error.code}, errorId=${errorId}, reason=soba)`,
+      `testing (errorCode="${error.code}", errorId="${errorId}", reason="soba")`,
     );
   });
 });
+
+describe('#publishEvent', () => {
+  test('should support publishEvent(...) function', async () => {
+    const logger = createIntegrationLogger({ name, invocationConfig });
+    const context: SynchronizationJobContext = {
+      logger,
+      job: { id: 'test-job-id' } as SynchronizationJob,
+      apiClient: createApiClient({
+        apiBaseUrl: 'https://api.us.jupiterone.io',
+        account: 'mocheronis',
+      }),
+    };
+
+    const expectedEventsUrl =
+      '/persister/synchronization/jobs/test-job-id/events';
+
+    logger.registerSynchronizationJobContext(context);
+
+    const postSpy = jest
+      .spyOn(context.apiClient, 'post')
+      .mockImplementation(noop as any);
+
+    logger.publishEvent({
+      name: 'the name',
+      description: 'the description',
+    });
+
+    await logger.flush();
+
+    expect(postSpy).toHaveBeenCalledTimes(1);
+    expect(postSpy).toHaveBeenNthCalledWith(1, expectedEventsUrl, {
+      events: [{ name: 'the name', description: 'the description' }],
+    });
+  });
+});
+
+describe('#publishErrorEvent', () => {
```

Comment:
Similar to `logger.warn` described elsewhere, we've been working toward having `logger.error` log an error, and then publish the error event (this already happens with thrown errors). This approach, along with the [other logger functions](https://github.com/JupiterOne/integration-sdk/blob/master/src/framework/execution/types/logger.ts#L45), is meant to help avoid random events in the job event log, but also provides a level of indirection so that when `j1-integration collect` is run, and there is [no synchronization job](https://github.com/JupiterOne/integration-sdk/blob/5932e1bee60e96b838aa8b372389518803c2393a/src/framework/execution/logger.ts#L192), `publishEvent` does nothing. That is, `logger.*` methods will output to the local log and, depending on the nature of their logging, `publishEvent` is used to record them in the job event log when one is available.

I see that you'd like to add additional data to the logged event for the user to see, and I suspect from previous conversations that this will be related to troubleshooting exceptional situations. I think we should consider having the error types (such as `IntegrationProviderAPIError` or whatever other error type we should create) receive the additional error context (the data) and instead of serializing it to the single line message, improve the data storage for the event to support storing that error object with its additional properties. Then we can show the error in the job log with the short message, and expand it to show the complete error.